vendor/engines/bulk_email/app/controllers/bulk_email/bulk_email_controller.rb

```diff
@@ -11,26 +11,64 @@ class BulkEmailController < ApplicationController
     before_action :authenticate_user!
     before_action :check_acting_as
     before_action :init_current_facility
+    before_action :init_delivery_form, only: [:create, :deliver]
     before_action { authorize! :send_bulk_emails, current_facility }
 
-    before_action :init_search_options
+    before_action :init_search_options, only: [:search]
 
+    helper_method :bulk_email_cancel_path
     helper_method :datepicker_field_input
     helper_method :user_type_selected?
 
     def search
-      @searcher = BulkEmailSearcher.new(@search_fields)
+      @searcher = RecipientSearcher.new(@search_fields)
       @users = @searcher.do_search
     end
 
-    def download
+    def create
       @users = User.where_ids_in(params[:recipient_ids])
-      filename = "bulk_email_recipients.csv"
-      set_csv_headers(filename)
+
+      respond_to do |format|
+        format.csv do
+          filename = "bulk_email_recipients.csv"
+          set_csv_headers(filename)
+        end
+
+        format.html
+      end
+    end
+
+    def deliver
+      if @delivery_form.deliver_all
+        flash[:notice] = text("bulk_email.delivery.success", count: @delivery_form.recipient_ids.count)
+        redirect_to facility_bulk_email_path
+      else
+        @users = User.where_ids_in(@delivery_form.recipient_ids)
+        render :create
+      end
     end
 
     private
 
+    def bulk_email_cancel_path
+      if cancel_params.present?
+        facility_bulk_email_path(cancel_params)
+      else
+        facility_bulk_email_path
+      end
+    end
+
+    def cancel_params
+      @cancel_params ||= params.select do |field, _value|
```

Comment:
Does `params.slice(%w(start_date end_date bulk_email products))` get what you want?